File: client/src/types/sale.ts
import { GetBaseWork } from "./base_work";

export interface CreateSaleDto {
    width: number;
    height: number;
    year: number;
    price: number;
    name_ru: string;
    name_en: string;
    base_id: number;
    materials_ids: number[];
    descr: string;
    images: string[];
}

export interface UpdateSaleResponse {
    id: number;
    str_id: string;
    dir: string;
    name_ru: string;
    name_en: string;
    base_id: number;
    year: number;
    descr: string;
    width: number;
    height: number;
    price: number;
    images: string[];
    materials_ids: number[];
}

export interface UpdateSaleRequest {
    id: number;
    name_ru: string;
    name_en: string;
    base_id: number;
    year: number;
    descr: string;
    width: number;
    height: number;
    price: number;
    images: string[];
    materials_ids: number[];
}

export type GetSaleDto = GetBaseWork & {
    price: number;
};

export type TypedGetSaleDto = GetSaleDto & {
    __type: "GetSaleDto";
};

File: server/dtos/sale.go
package dtos

import (
	_ "github.com/mattn/go-sqlite3"
)

type SaleResponse struct {
	Id          int      `json:"id"`
	StrId       string   `json:"str_id"`
	Dir         string   `json:"dir"`
	NameRu      string   `json:"name_ru"`
	NameEn      string   `json:"name_en"`
	Year        int      `json:"year"`
	DescrRu     string   `json:"descr_ru"`
	DescrEn     string   `json:"descr_en"`
	BaseRu      string   `json:"base_ru"`
	BaseEn      string   `json:"base_en"`
	Width       int      `json:"width"`
	Height      int      `json:"height"`
	Price       int      `json:"price"`
	Images      []string `json:"images"`
	MaterialsEn []string `json:"materials_en"`
	MaterialsRu []string `json:"materials_ru"`
}

// Служит для отображения информации о картине на продажу на странице редактирования.
type UpdateSaleResponse struct {
	Id           int      `json:"id"`
	StrId        string   `json:"str_id"`
	Dir          string   `json:"dir"`
	NameRu       string   `json:"name_ru"`
	NameEn       string   `json:"name_en"`
	BaseId       int      `json:"base_id"`
	Year         int      `json:"year"`
	DescrRu      string   `json:"descr_ru"`
	DescrEn      string   `json:"descr_en"`
	Width        int      `json:"width"`
	Height       int      `json:"height"`
	Price        int      `json:"price"`
	Images       []string `json:"images"`
	MaterialsIds []int    `json:"materials_ids"`
}

// UpdateSaleRequest is the DTO for updating a sale from frontend.
type UpdateSaleRequest struct {
	Id           int      `json:"id"`
	NameRu       string   `json:"name_ru"`
	NameEn       string   `json:"name_en"`
	BaseId       int      `json:"base_id"`
	Year         int      `json:"year"`
	DescrRu      string   `json:"descr_ru"`
	DescrEn      string   `json:"descr_en"`
	Width        int      `json:"width"`
	Height       int      `json:"height"`
	Price        int      `json:"price"`
	Images       []string `json:"images"`
	MaterialsIds []int    `json:"materials_ids"`
}

// CreateSaleRequest is the DTO for creating a sale from frontend.
type CreateSaleRequest struct {
	NameRu       string   `json:"name_ru"`
	NameEn       string   `json:"name_en"`
	BaseId       int      `json:"base_id"`
	Year         int      `json:"year"`
	DescrRu      string   `json:"descr_ru"`
	DescrEn      string   `json:"descr_en"`
	Width        int      `json:"width"`
	Height       int      `json:"height"`
	Price        int      `json:"price"`
	Images       []string `json:"images"`
	MaterialsIds []int    `json:"materials_ids"`
}
